src/app/shared/context.service.ts

```diff
@@ -51,107 +46,123 @@ export class ContextService implements Contexts {
-          if (val && val.id) {
-            return {
-              user: null,
-              space: val,
-              type: ContextTypes.BUILTIN.get('user'),
-              name: 'TO_DELETE',
-              path: null
-            } as Context;
-          } else {
-            return {} as Context;
-          }
-        })
-      ).subscribe(val => {
-        this._deleteFromRecent.next(val);
-      });
-    // Initialize the default context when the logged in user changes
-    this._default = this.userService.loggedInUser
-      .pipe(
-      // First use map to convert the broadcast event to just a username
-        map(val => {
-          if (!(val && val.id)) {
-            // this is a logout event
-          } else if (val.attributes.username) {
-            this._currentUser = val.attributes.username;
-            return val.attributes.username;
-          } else {
-            this.notifications.message({
-              message: 'Something went badly wrong. Please try again later or ask for help.',
-              type: NotificationType.DANGER
-            } as Notification);
-            throw 'Unknown user';
-          }
-        }),
-        distinctUntilChanged(),
-        // Then, perform another map to create a context from the user
-        switchMap(val => this.userService.getUserByUsername(val)),
-        map(val => {
-          if (val && val.id) {
-            return {
-              user: val,
-              space: null,
-              type: ContextTypes.BUILTIN.get('user'),
-              name: val.attributes.username,
-              path: '/' + val.attributes.username
-            } as Context;
-          } else {
-            return {} as Context;
-          }
-        })
-      ).multicast(() => new ReplaySubject(1));
+    private toggleService: FeatureTogglesService
+  ) {
+    this._recent = new ReplaySubject<Context[]>(1);
+    this.initDefault();
+    this.initRecent();
+  }
 
-    // Create the recent space list
-    this._recent = merge(this._addRecent, this._deleteFromRecent).pipe(
-      // Map from the context being added to an array of recent contexts
-      // The scan operator allows us to access the list of recent contexts and add ours
-      scan(this.updateRecentSpaceList, []),  // The final value to scan is the initial value, used when the app starts
-      // Finally save the list of recent contexts
-      tap(val => {
-        // Truncate the number of recent contexts to the correct length
-        if (val.length > this.RECENT_CONTEXT_LENGTH) {
-          val.splice(
-            this.RECENT_CONTEXT_LENGTH,
-            val.length - this.RECENT_CONTEXT_LENGTH
-          );
+  initDefault(): void {
+   // Initialize the default context when the logged in user changes
+   this._default = this.userService.loggedInUser
+    .pipe(
+    // First use map to convert the broadcast event to just a username
+      map((val: User): string => {
+        if (!(val && val.id)) {
+          // this is a logout event
```

Comment:
Should there be an explicit return here?

Reply:
Yeah it'd probably be best to return something here. Since there is no valid username at this point, I've made it return a null that gets filtered out before continuing with the function.